lib/auth/auth.go

```diff
@@ -1170,6 +1170,88 @@ func (a *Server) GetMFADevices(ctx context.Context, req *proto.GetMFADevicesRequ
 	}, nil
 }
 
+// verifyMFARespAndAddDeviceWithToken validates MFA register response and on success,
+// adds the new MFA device. This method is specifically for secrets and storage keys associated with a user token.
+func (a *Server) verifyMFARespAndAddDeviceWithToken(ctx context.Context, regResp *proto.MFARegisterResponse, token types.UserToken, deviceName string) error {
+	cap, err := a.GetAuthPreference(ctx)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	// Validate MFARegisterResponse and upsert the new device on success.
+	var dev *types.MFADevice
+	switch resp := regResp.GetResponse().(type) {
```

Comment:
We need to conciliate this switch with the one in [grpcserver.go/addMFADeviceRegisterChallenge](https://github.com/gravitational/teleport/blob/f825d4558c1de466957577d4d973589a85d49077/lib/auth/grpcserver.go#L1734). Please either do that or leave a TODO for me, I really don't want to have this logic sprawling around.

Reply:
I gave it a shot to combine it for different use case. There are some things to note:

- When `tsh` users try to add a device that is not supported by cluster, user now gets the following errors (whereas before, there were none):
  - second factor U2F not allowed by cluster or
  - second factor TOTP not allowed by cluster
- I did not check for this error, I didn't think it was too big a deal to try and handle this error:
```
challenge := regChallenge.GetTOTP()
		if challenge == nil {
			return nil, trace.BadParameter("got unexpected %T in response to %T", regResp.Response, regChallenge.Request)
		}
```

Reply:
Both seem reasonable to me. Thanks for taking a shot at this.